clang/lib/CodeGen/CodeGenTypes.cpp

```diff
@@ -808,7 +808,7 @@ llvm::Type *CodeGenTypes::ConvertType(QualType T) {
       SkippedLayout = true;
       ResultType = llvm::Type::getInt8Ty(getLLVMContext());
     }
-    ResultType = llvm::ArrayType::get(ResultType, 0);
```

Comment:
Can you explain how this change helps you? I am not sure what the implications of setting array size to 1 here are but I would assume it is incorrect. Does it even represent an incomplete array if we set the size to 1?  @tahonermann @premanandrao please weigh in here.

Reply:
I'm guessing this is related to the outstanding issue from the PR description: "For some reason, the code may fail to compile when using unbounded arrays due to a LLVM IR -> SPIRV translator problem". Presumably this will be addressed before this PR is merged? If not, a FIXME comment is definitely warranted.

Reply:
Note that I've made a recent change to only do this if we are compiling for a SYCL device.
The issue is that a zero-sized array is forbidden in SPIR-V so this essentially does not allow us to use unbounded arrays in SYCL kernels which this extension demands. This is what this "hack" is trying to circumvent.

Reply:
To elaborate a bit more, when using unbounded arrays inside SYCL kernels, we are encountering a SPIRV translator error. The unbounded array is lowered into the llvm type `[0 x element_type]` and unfortunately when this is further compiled into SPIR-V, it rejects it because zero-sized arrays are not allowed in SPIR-V. See https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpTypeArray

Reply:
That sounds like a defect in the LLVM IR -> SPIR-V translation or a gap in the SPIR-V specification itself. Unbounded arrays are a standard C++ language feature. How is this handled elsewhere in code gen? Perhaps these cases should be lowered to a pointer rather than to the LLVM 0-sized array type. Or perhaps the LLVM 0-sized array type should be lowered to a pointer in SPIR-V.

Reply:
The revert is good. Handling the unbounded array as an array of size one might still be problematic. I'm not familiar with SPIR-V tooling, but I can imagine memory sanitizer tools to potentially flag an access to elements beyond element 0 as an out-of-bounds read/write. This must already be handled in some way. Consider the IR generated for the following example; note the `[0 x i32]` used with `getelementptr`. https://godbolt.org/z/bjh5Ev6ah.
```
int f(int(&a)[]) {
  return a[2];
}
void g() {
  int a[] = { 0, 1, 2, 3 };
  f(a);
}
```

Reply:
>>That sounds like a defect in the LLVM IR -> SPIR-V translation or a gap in the SPIR-V specification itself. 

I agree with Tom here. Handling an unbounded array as an array of size 1 seems like a hack irrespective of whether this is done in AST->LLVM IR lowering or LLVMIR->SPIRV lowering.

Reply:
I understand, tagging @MrSidims on whether we can make this hack work or if it just breaks existing code.

Reply:
> This must already be handled in some way. Consider the IR generated for the following example; note the `[0 x i32]` used with `getelementptr`. https://godbolt.org/z/bjh5Ev6ah.

I think there is no support currently for unbounded arrays, and that is the fundamental problem.  The code snippet you show also does not compile to SPIR-V, producing the same error in the SPIRV translator that @lbushi25 saw earlier.

BTW, thanks for the godbolt link.  I didn't realize this cast from `int[4]` to `int[]` was allowed without a `reinpterpret_cast`.  It seems that you need the `reinterpret_cast` in C++17 mode, but the cast is implicit in C++20 mode.

Reply:
After giving it some more thought, I think this change is probably fine. Calling it a hack in my opinion is not quite right, the LLVM IR reference for example says that the way that they choose to represent unknown-sized arrays as arrays of length zero in IR is merely a convention and that there's nothing special about the number zero. 

In SPIR-V, the idiomatic way to access an array is using OpAccessChain and OpLoad, neither of which seem to do bounds-checking at compile-time, rather you get undefined behavior if an index is out of bounds so I think this allows us extend that LLVM IR way of handling unknown-sized arrays except that the convention in this case should be 1-length arrays since zero-length arrays are not supported. I'm unfamiliar with SPIR-V, so I'd like to know if @MrSidims agrees with this rationale.

Reply:
@gmlueck, implicit conversions were added in C++20 courtesy of [P0388R4 (Permit conversions to arrays of unknown bound)](https://wg21.link/p0388r4).

@lbushi25, thanks for that link. I'm also unfamiliar with SPIR-V, so I'm curious if these questions correspond to a gap in the specification. Unless it is actually clear somewhere in the SPIR-V specification (I wasn't able to identify wording that made things sufficiently clear to me), I think it would be worth filing a new issue against the SPIR-V specification.